<a href="https://colab.research.google.com/github/Ripperonnay/Machine-Learning-Projects/blob/main/SONAR_Rock_vs_Mine_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**To create a model that predicts if the object is a mine or a rock when Sonar sends sound waves reflected back**

**WORK FLOW**

Sonar Data -> Pre-Processing -> Train Test (Split the data into training cases and Testing cases)
We'll use a Logistic Regression Model because it is really good for binary classification

In [ ]:
#dependencies
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Data Collection and Data Processing

In [ ]:
#Loading the dataset to a pandas Dataframe
sonar_data = pd.read_csv('/content/sonar data.csv', header = None)  #we put header = None as we don't have any column names available
sonar_data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [ ]:
sonar_data.shape    #rows and columns

(208, 61)

In [ ]:
sonar_data.describe()   #shows all the stats for all columns

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [ ]:
sonar_data[60].value_counts()       #Shows how many rocks and how many miles, basically shows total count of distinct values in 60th column
                                    #Return a Series containing counts of unique values.

M    111
R     97
Name: 60, dtype: int64

M --> Mine
R --> Rock

In [ ]:
sonar_data.groupby(60).mean()        #This finds the mean value for all the rows for M and R
#A groupby operation involves some combination of splitting the object, applying a function,
#and combining the results. This can be used to group large amounts of data and compute operations on these groups.

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [ ]:
#Now we have to seperate the data and the Labels = M,R
X = sonar_data.drop(columns=60, axis = 1)        #this drops the last column from the dataframe, axis=1 specifies that we are dropping a colunmn
Y = sonar_data[60]                               #we are storing the labels in Y

In [ ]:
print(X.head())
print(Y.head())

       0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.016

Splitting training and test data

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size = 0.1, stratify = Y, random_state = 1)
#this order is necessary and should be maintained

train_test_split : Split arrays or matrices into random train and test subsets.

test_size = 0.1 : represents the proportion of which should be test data, 10% in this case (should be 0.1 or 0.2)

stratify = Y : we need to have data to be split so that the no of rocks dataset and no of mines data set are almost equal in the training/test dataset. (what does it do exactly tho?)

random_state = 1 : splits the data in a particular order, so if 2 people put random_state = 1, there data will be split similarly.

In [ ]:
print (X.shape, X_train.shape, X_test.shape)

(208, 60) (187, 60) (21, 60)


X of 208 records split into training = 187 and test = 21

**Now we will train our Logistic Regression MODEL**

In [ ]:
#creating the model
model = LogisticRegression()

In [ ]:
#Training the model
model.fit(X_train, Y_train)

LogisticRegression()

**MODEL EVALUATION**

Now we will evaluate the accuracy of our model. How?

We will run the model on the data and match the results. Any accuracy >70% is good

In [ ]:
#accuracy for training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [ ]:
training_data_accuracy*100

83.42245989304813

In [ ]:
#accuracy for test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [ ]:
test_data_accuracy*100     #76%

76.19047619047619

Since now accuracy is good, we will make a **Predictive System**. It will predict given a input data set, whether the object is a rock or a mine

In [ ]:
input_data = [0.0305,0.0363,0.0214,0.0227,0.0456,0.0665,0.0939,0.0972,0.2535,0.3127,0.2192,0.2621,0.2419,0.2179,0.1159,0.1237,0.0886,0.1755,0.1758,0.1540,0.0512,0.1805,0.4039,0.5697,0.6577,0.7474,0.8543,0.9085,0.8668,0.8892,0.9065,0.8522,0.7204,0.6200,0.6253,0.6848,0.7337,0.6281,0.5725,0.6119,0.5597,0.4965,0.5027,0.5772,0.5907,0.4803,0.3877,0.2779,0.1427,0.0424,0.0271,0.0200,0.0070,0.0070,0.0086,0.0089,0.0074,0.0042,0.0055,0.0021]
#changing the input data to a numpy array
input_data_np = np.asarray(input_data).reshape(1,-1)         #reshaping to 1 data record
prediction = model.predict(input_data_np)
print("Rock" if prediction[0] == 'R' else "Mine")

Mine


COOL SHIT

**All steps**


1. import libraries, dependencies
2. load the dataset to a pandas Dataframe
3. check shape, stats, etc to get overview of the data
4. seperate the data and the label
5. split training data and testing data using train_test_split
6. create the ML Model (Logistic Regression in this case)
7. fit the training data and label to the model
8. check accuracy (training, testing)
7. make a predictive system


